### DISCUSSION TITLE
Is it possible to specify ORM relationship join conditions at query time?

### DISCUSSION THREAD

**Discussion by synic** [sqlalchemy/sqlalchemy#6932] [Usage Questions] [answered]
Suppose I have the following models:

```python

class Integration(BaseModel):
    name = sa.Column(sa.Unicode(length=100))
    company_integration = relationship('CompanyIntegration', lazy='noload')

    __tablename__ = 'integrations'

class CompanyIntegration(BaseModel):
    company_id = sa.Column(UUID, sa.ForeignKey('companies.id'), nullable=False)
    company = relationship('Company', lazy='raise')

    integration_id = sa.Column(UUID, sa.ForeignKey('integrations.id'), nullable=False)
    integration = relationship(Integration, lazy='raise')

    settings = sa.Column(JSONB, default=sa.text("'{}'"), nullable=False)
    is_enabled = sa.Column(sa.Boolean, default=False, nullable=False)

    __tablename__ = 'company_integrations'
```

In this case, `Integration` is an available integration in the system, for example, "Slack". A `CompanyIntegration` is created when a Company enables an integration. It contains settings for that integration, for that company.

If I know the company, I would like to populate `Integration.company_integration` with the `CompanyIntegration` for that company. If I do not know the company, then I would like it to stay deferred.

Is there a way to dynamically populate `relationship` like that at query time? Does this question even make sense?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I think you can use the normal [eager loading](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html), If an object has no `company_integration ` it will not be deferred but be loaded as an empty list. Would that work?

**Reply by synic** [reply]
It's actually a one-to-one, but I was able to make it work by setting `lazy='noload'` on the `company_integration` relationship, and querying like so:

```python
stmt = (
    sa.select(Integration)
    .outerjoin(
        CompanyIntegration,
        sa.and_(
            CompanyIntegration.integration.id == Integration.id,
            CompanyIntegration.company_id == company_id,
        ),
    )
    .options(contains_eager(Integration.company_integration))
)
```

Does that seem right? Then, if I don't know `company_id`, I can leave the `options` part out and `integration.company_integration` will be `None`. When I do know the `company_id`, if a `CompanyIntegration` exists, it will be populated in `company_integration`. If one doesn't exist, `company_integration` will be None.

**Reply by CaselIT** [reply]
> It's actually a one-to-one

Then I think you should specify `uselist=False` in the `company_integration ` relationship, so you have a single object or `None` instead of a list.

That query seems to be ok, but it's probably just easier to use an eager loading, like selectinload

**Reply by synic** [reply]
I guess I am confused about how to limit the `CompanyIntegration` by `company_id` (I only want the `CompanyIntegration` that matches `company_id`, or `None` if one doesn't exist) with a regular joinedload or selectinload.  

Basically, I want _all_ integrations, and for `integration.company_integration` to be populated with the `CompanyIntegration` that matches `company_id`, if one exists, otherwise, it should be `None`.

**Reply by CaselIT** [reply]
Oh sorry I missed that. You can use this https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#loader-option-criteria

**Reply by synic** [reply]
Ooh, that's much easier. Thank you!
